File: Kakao/2020_자물쇠와 열쇠.cpp
// 출제의도: "구현", Time Complexity: O((M+N)^2*(M^2+N^2)), Space Complexty: O((M+N)^2).

#include <string>
#include <vector>
#include <iostream>

using namespace std;

typedef pair<int,int>   Pos;
int M, N;

void rotate_clockwise_90(vector<vector<int>> &key) {
    vector<vector<int>> key_rotated = key;
    for(int i=0;i<M;i++) {
        for(int j=0;j<M;j++) {
            key_rotated[j][M-1-i] = key[i][j]; 
        }
    }
    key = key_rotated;
}

void map_key(vector<vector<int>> &key_union_lock, vector<vector<int>> &key, Pos left_top) {
    for(int i=0;i<M;i++) {
        for(int j=0;j<M;j++) {
            key_union_lock[left_top.first+i][left_top.second+j] += key[i][j];
        }
    }
}

void map_lock(vector<vector<int>> &key_union_lock, vector<vector<int>> &lock) {
    for(int i=M-1;i<=N+M-2;i++) {
        for(int j=M-1;j<=N+M-2;j++) {
            key_union_lock[i][j] = lock[i-(M-1)][j-(M-1)];
        }
    }
}

bool check_key_fits_lock(vector<vector<int>> &key_union_lock) {
    for(int i=M-1;i<=N+M-2;i++) {
        for(int j=M-1;j<=N+M-2;j++) {
            if(key_union_lock[i][j]!=1)   return false;
        }
    }
    return true;
}

bool insert_key_into_lock(vector<vector<int>> &key, vector<vector<int>> &lock) {
    for(int i=0;i<=N+M-2;i++) {
        for(int j=0;j<=N+M-2;j++) {
            vector<vector<int>> key_union_lock(N+2*M-2,vector<int>(N+2*M-2,0));
            map_lock(key_union_lock, lock);
            map_key(key_union_lock,key,Pos(i,j));
            if(check_key_fits_lock(key_union_lock))   return true;
        }
    }
    return  false;
}

bool solution(vector<vector<int>> key, vector<vector<int>> lock) {
    bool answer = false;
    M = key.size();
    N = lock.size();
    for(int d=90;d<=360;d+=90) {
        rotate_clockwise_90(key);
        if(insert_key_into_lock(key,lock)) {
            answer = true;
            break;
        }
    }
    return answer;
}

File: Kakao/2020_자물쇠와 열쇠.swift
// 출제의도: "구현"
// c++에 비교하여 swift 언어 사용의 장점은 없었던 문제.

import Foundation

func rotateClockwise90(_ key: [[Int]]) -> [[Int]] {
    let m = key.count
    var keyRotated = Array(repeating: Array(repeating: 0, count: m), count: m)
    
    for i in 0..<m {
        for j in 0..<m {
            keyRotated[j][m-i-1] = key[i][j]
        }
    }
    return keyRotated
}

func putLockOnBoard(_ board: inout [[Int]], _ lock: [[Int]], _ key: [[Int]]) -> Void {
    let m = key.count, n = lock.count
    
    for i in m-1..<m+n-1 {
        for j in m-1..<m+n-1 {
            board[i][j] += lock[i-(m-1)][j-(m-1)]
        }
    }
}

func putKeyOnBoard(_ board: inout [[Int]], _ key: [[Int]], _ pos: [Int]) -> Void {
    let m = key.count
    
    for i in pos[0]..<pos[0]+m {
        for j in pos[1]..<pos[1]+m {
            board[i][j] += key[i-pos[0]][j-pos[1]]
        }
    }
}

func keyFitsLock(_ board: [[Int]], _ lock: [[Int]], _ key: [[Int]]) -> Bool {
    let m = key.count, n = lock.count
    
    for i in m-1..<m+n-1 {
        for j in m-1..<m+n-1 {
            if board[i][j] != 1 {
                return false
            }
        }
    }
    return true
}

func insertKeyIntoLock(_ lock: [[Int]], _ key: [[Int]]) -> Bool {
    let m = key.count, n = lock.count
    
    for i in 0..<m+n-1 {
        for j in 0..<m+n-1 {
            var board = Array(repeating: Array(repeating: 0, count: 2*m+n-2), count: 2*m+n-2)
            putLockOnBoard(&board, lock, key)
            putKeyOnBoard(&board, key, [i,j])
            if(keyFitsLock(board, lock, key)) {
                return true
            }
        }
    }
    return false
}

func solution(_ key:[[Int]], _ lock:[[Int]]) -> Bool {
    var key = key
    for i in 0..<4 {
        if(insertKeyIntoLock(lock, key)) {
            return true
        }
        key = rotateClockwise90(key)
    }
    return false
}
